openshift-auth-proxy.js

```diff
@@ -17,89 +17,168 @@ var argv = require('yargs')
+      demand: true,
+      default: process.env.OAP_PUBLIC_MASTER_URL
+    }, 'master-ca': {
+      describe: 'CA certificate(s) file to validate connection to the master',
+      demand: true,
+      default: process.env.OAP_MASTER_CA_FILE || 'secret/master-ca'
+    }, 'proxy-cert': {
       describe: 'Certificate file to use to listen for TLS',
-      demand: true
-    }, 'tls-key': {
+      demand: true,
+      default: process.env.OAP_PROXY_CERT_FILE || 'secret/proxy-cert'
+    }, 'proxy-key': {
       describe: 'Key file to use to listen for TLS',
-      demand: true
+      demand: true,
+      default: process.env.OAP_PROXY_KEY_FILE || 'secret/proxy-key'
+    }, 'proxy-tlsopts-file': {
+      describe: 'File containing JSON for proxy TLS options',
+      demand: true,
+      default: process.env.OAP_PROXY_TLS_FILE || 'secret/proxy-tls.json'
+    }, 'mutual-tls-ca': {
+      describe: 'CA cert file to use for validating TLS client certs under "mutual_tls" auth method',
+      demand: false,
+      default: process.env.OAP_PROXY_CA_FILE || 'secret/proxy-ca'
     }
   })
   .help('help')
   .epilog('copyright 2015')
   .argv;
 
-if (argv['session-secret'] === 'generated') {
-  require('base64url')(require('crypto').randomBytes(256)).substring(0, 256);
+// ---------------------- config --------------------------
+
+//
+// read in all the files with secrets, keys, certs
+//
+var sessionSecret;
+try {
+  sessionSecret = fs.readFileSync(argv['session-secret'], "utf8");
+} catch(err) {
+  console.error("error reading session secret: %s", JSON.stringify(e));
+} finally { // just ignore if the file is not there
+  if (sessionSecret == null ) {
+    console.error("generating session secret (will not work with scaled service)");
+    sessionSecret = require('base64url')(require('crypto').randomBytes(256)).substring(0, 256);
+  }
+};
+var clientSecret = fs.readFileSync(argv['client-secret'], "utf8").replace(/(\n|\r)/gm,"");
+var masterCA = fs.readFileSync(argv['master-ca'], "utf8");
+var mutualTlsCa;
+try { // it's optional...
+  mutualTlsCa = fs.readFileSync(argv['mutual-tls-ca'], "utf8");
+} catch(err) {
+  if (argv['auth-mode'] == 'mutual_tls') {
+    throw "No CA read for mutual TLS. Looked in: " + argv['mutual-tls-ca'];
+  } // otherwise, we don't need it
+};
+var proxyTLS = {};
+try { // also optional TLS overrides (ciphersuite etc)
+  proxyTLS = fs.readFileSync(argv['proxy-tlsopts-file'], "utf8");
+  if (argv.debug) { console.log("Read TLS opts from %s: %s", argv['proxy-tlsopts-file'], proxyTLS); }
+  proxyTLS = eval(proxyTLS);
+  if (proxyTLS == null || ! typeof proxyTLS === 'object') {
+    throw("TLS opts file did not evaluate to an object");
+  }
+} catch(e) {
+  console.error("Could not read TLS opts from %s; error was: %s", argv['proxy-tlsopts-file'], e);
+  proxyTLS = {};
+} finally {
+  proxyTLS['key'] = fs.readFileSync(argv['proxy-key'], "utf8");
+  proxyTLS['cert'] = fs.readFileSync(argv['proxy-cert'], "utf8");
```

Comment:
I should supply some in the example... our security guys had some pretty specific indications of what they wanted for OpenShift online, could start with those.